uaclient/data_types.py

```diff
@@ -0,0 +1,133 @@
+            INCORRECT_TYPE_ERROR_MESSAGE.format(
+                type=expected_type, value=repr(got_value)
+            )
+        )
+
+
+class IncorrectListElementTypeError(IncorrectTypeError):
+    def __init__(self, err: IncorrectTypeError, at_index: int):
+        self.msg = INCORRECT_LIST_ELEMENT_TYPE_ERROR_MESSAGE.format(
+            index=at_index, nested_msg=err.msg
+        )
+
+
+class IncorrectFieldTypeError(IncorrectTypeError):
+    def __init__(self, err: IncorrectTypeError, key: str):
+        self.msg = INCORRECT_FIELD_TYPE_ERROR_MESSAGE.format(
+            key=key, nested_msg=err.msg
+        )
+
+
+class DataValue:
+    @staticmethod
+    def from_value(val: Any) -> Any:
+        return val
+
+
+class StringDataValue(DataValue):
+    @staticmethod
+    def from_value(val: Any) -> str:
+        if not isinstance(val, str):
+            raise IncorrectTypeError("string", val)
+        return val
+
+
+class IntDataValue(DataValue):
+    @staticmethod
+    def from_value(val: Any) -> int:
+        if not isinstance(val, int):
+            raise IncorrectTypeError("int", val)
+        return val
+
+
+def data_list(data_cls: Type[DataValue]) -> Type[DataValue]:
+    class _DataList(DataValue):
+        @staticmethod
+        def from_value(val: Any) -> List:
+            if not isinstance(val, list):
+                raise IncorrectTypeError("list", val)
+            ret_list = []
+            for i, item in enumerate(val):
+                try:
+                    ret_list.append(data_cls.from_value(item))
+                except IncorrectTypeError as e:
+                    raise IncorrectListElementTypeError(e, i)
+            return ret_list
+
+    return _DataList
+
+
+class Field:
+    def __init__(
+        self, key: str, data_cls: Type[DataValue], required: bool = True
+    ):
+        self.key = key
+        self.data_cls = data_cls
+        self.required = required
+
+
+T = TypeVar("T", bound="DataObject")
+
+
+class DataObject(DataValue):
+    fields = []  # type: List[Field]
+
+    def __init__(self, **_kwargs):
+        pass
+
+    @classmethod
+    def from_dict(cls: Type[T], d: dict) -> T:
+        kwargs = {}
+        for field in cls.fields:
+            try:
+                val = d[field.key]
+            except KeyError:
+                if field.required:
+                    raise IncorrectFieldTypeError(
+                        IncorrectTypeError(field.data_cls.__name__, None),
+                        field.key,
+                    )
+                else:
+                    val = None
+            if val is not None:
+                try:
+                    val = field.data_cls.from_value(val)
+                except IncorrectTypeError as e:
+                    raise IncorrectFieldTypeError(e, field.key)
+            kwargs[field.key] = val
+        return cls(**kwargs)
+
+    @classmethod
+    def from_value(cls, val: Any):
+        if not isinstance(val, dict):
+            raise ValueError()
```

Comment:
Why not use the `IncorrectTypeError` exception here ?

Reply:
Ah good catch! I just missed this one when I was adding the error functionality. I'll fix it